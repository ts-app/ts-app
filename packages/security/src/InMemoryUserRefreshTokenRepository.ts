import { of } from 'rxjs'
import { UserRefreshTokenRepository } from './UserRefreshTokenRepository'

export class InMemoryUserRefreshTokenRepository extends UserRefreshTokenRepository {
  tokens: {
    [ key: string ]: Set<string>
  } = {}

  register (userId: string, refreshToken: string) {
    if (!this.tokens[ userId ]) {
      this.tokens[ userId ] = new Set()
    }

    this.tokens[ userId ].add(refreshToken)
    return of(null)
  }

  getTokens (userId: string) {
    if (!this.tokens[ userId ]) {
      this.tokens[ userId ] = new Set()
    }

    return of(this.tokens[ userId ])
  }

  clearTokens (userId?: string) {
    if (!userId) {
      this.tokens = {}
    } else {
      const userTokens = this.tokens[ userId ]
      if (userTokens) {
        userTokens.clear()
      }
    }
    return of(null)
  }
}
