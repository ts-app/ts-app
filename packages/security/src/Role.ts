import { Timestampable } from '@ts-app/common'

export interface Role extends Timestampable {
  id: string
  name: string
}
