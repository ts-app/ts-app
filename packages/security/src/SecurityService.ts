import { Observable } from 'rxjs'
import { User, UserProfile, BusinessError, FindInput, FindOutput } from '@ts-app/common'

export abstract class SecurityService {
  abstract signUp (input: { email: string, password: string }): Observable<{
    user?: User
    error?: BusinessError
  }>

  abstract loginWithEmailPassword (input: { email: string, password: string }): Observable<{
    user?: User
    accessToken?: string
    refreshToken?: string
    error?: BusinessError
  }>

  abstract loginWithRefreshToken (refreshToken: string): Observable<{
    user?: User
    accessToken?: string
    refreshToken?: string
  }>

  abstract user (id: string): Observable<User | null>

  abstract users (input: FindInput): Observable<FindOutput<User>>

  abstract userByEmail (email: string): Observable<User | null>

  abstract removeUser (id: string): Observable<null>

  abstract seedUsers (input: {
    force?: boolean
    userCount?: number
  }): Observable<null>

  abstract updateProfile<T extends UserProfile> (input: {
    id: string
    profile: Partial<T>
  }): Observable<null>

  abstract reset (): Observable<null>
}
