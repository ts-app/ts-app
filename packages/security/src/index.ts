export * from './MongoRoleService'
export * from './MongoSecurityService'
export * from './RoleService'
export * from './SecurityService'
