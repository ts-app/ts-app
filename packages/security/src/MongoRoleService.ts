import { MongoService } from '@ts-app/mongo'
import { assert, FindInput, FindOutput, User, makeTimestampable } from '@ts-app/common'
import { Observable, of, throwError, from } from 'rxjs'
import { concatMap, mapTo, tap, map, toArray, defaultIfEmpty } from 'rxjs/operators'
import { RoleService } from './RoleService'
import { Role } from './Role'
import { ObjectId } from 'bson'

export class MongoRoleService extends RoleService {
  constructor (private mongoService: MongoService) {
    super()
  }

  addUsersToRoles (input: { userIds: string[]; roles: string[]; group?: string }): Observable<null> {
    input.group = input.group || MongoRoleService.GLOBAL.replace(/\./g, '_')

    if (input.group.startsWith('$')) {
      return throwError('Group cannot start with \'$\'')
    }

    // --- create missing roles in role collection
    return this.mongoService.cursor<Role>('roles', { name: { $in: input.roles } }).pipe(
      concatMap(cursor => cursor.toArray()),
      concatMap(matchingRolesInDb => {
        const rolesNotInDb = input.roles.filter(roleName => !matchingRolesInDb.find(roleInDb => roleInDb.name === roleName))
        return of(...rolesNotInDb).pipe(
          concatMap(roleName => this.createRole(roleName))
        )
      }),

      // --- assign roles to users collection
      concatMap(() => this.mongoService.collection('users')),

      concatMap(Users => Users.updateMany(
        {
          _id: { $in: input.userIds.map(id => new ObjectId(id)) }
        },
        {
          $addToSet: {
            roles: {
              $each: input.roles.map(role => ({ role, group: input.group }))
            }
          }
        })),
      mapTo(null)
    )
  }

  createRole (name: string): Observable<string> {
    assert(!!name, 'Role name required')

    return this.mongoService.create<Role>(
      'roles',
      { name, ...makeTimestampable() }
    )
  }

  findRoles (input: FindInput): Observable<FindOutput<Role>> {
    return this.mongoService.find('roles', input, [ 'name' ])
  }

  getGroupsForUser (input: { userId: string; role?: string }): Observable<string[]> {
    const { role, userId } = input
    let filter
    if (role) {
      filter = {
        _id: new ObjectId(userId),
        'roles.role': role
      }
    } else {
      filter = { _id: new ObjectId(userId) }
    }

    return this.mongoService.get<User>('users', filter, {
      fields: { roles: 1 }
    }).pipe(
      concatMap(user => {
        if (!user) {
          // error if cannot find user, return empty array
          return of([])
        } else if (role) {
          // return groups where user has specified role assigned (except global)
          const groupNames = user.roles
            .filter(currentRole => currentRole.group !== RoleService.GLOBAL && currentRole.role === role)
            .map(currentRole => currentRole.group)

          return of(Array.from(new Set(groupNames)))
        } else {
          // role not specified, return all groups assigned to user (except global)
          const groupNames = user.roles.filter(currentRole => currentRole.group !== RoleService.GLOBAL)
            .map(currentRole => currentRole.group)

          return of(Array.from(new Set(groupNames)))
        }
      })
    )
  }

  getRolesForUser (input: { userId: string; group?: string }): Observable<string[]> {
    const { userId, group } = input
    let filter: any = { _id: new ObjectId(userId) }

    if (group) {
      filter = {
        ...filter,
        'roles.group': { $in: [ RoleService.GLOBAL, group ] }
      }
    }

    return this.mongoService.get<User>('users', filter, { fields: { roles: 1 } }).pipe(
      map(user => {
        if (user && user.roles) {
          // filter for global group
          // if group specified, filter by group
          const roles = user.roles
            .filter(role => role.group === RoleService.GLOBAL || !group || role.group === group)
            .map(role => role.role)

          return Array.from(new Set(roles))
        } else {
          return []
        }
      })
    )
  }

  getUsersInRoles (input: { roles: string[]; group?: string; limit?: number; cursor?: string }): Observable<FindOutput<User>> {
    return undefined
  }

  isUserInRoles (input: { userId: string; roles: string[]; group?: string }): Observable<boolean> {
    return undefined
  }

  removeRole (name: string): Observable<boolean> {
    return this.mongoService.get(
      'users',
      { 'roles.role': name },
      { fields: { _id: 1 } }
    ).pipe(
      concatMap(userWithRole => userWithRole ? of(false) :
        this.mongoService.remove('roles', { name }).pipe(
          map(remove => !!remove)
        ))
    )
  }

  removeUsersFromAllRoles (input: { userIds: string[] }): Observable<null> {
    return undefined
  }

  removeUsersFromRoles (input: { userIds: string[]; roles: string[]; group?: string }): Observable<null> {
    return undefined
  }

}
