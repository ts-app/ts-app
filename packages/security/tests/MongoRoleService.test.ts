import { ConsoleLogService, LogService, omitVolatile, User } from '@ts-app/common'
import { MongoService } from '@ts-app/mongo'
import { of } from 'rxjs'
import { catchError, concatMap, tap, toArray, mapTo, map } from 'rxjs/operators'
import { MongoRoleService, RoleService, MongoSecurityService, SecurityService } from '../src'

describe('MongoRoleService', async () => {
  const localUrl = 'mongodb://localhost:27017'
  let logService: LogService = new ConsoleLogService()
  let mongoService: MongoService
  let securityService: SecurityService
  let roleService: RoleService

  beforeEach(done => {

    mongoService = new MongoService(localUrl)
    securityService = new MongoSecurityService(mongoService)
    roleService = new MongoRoleService(mongoService)
    of('users', 'roles').pipe(
      concatMap(collection => mongoService.dropCollection(collection)),
      catchError(e => {
        // just ignore if cannot drop
        console.debug(e)
        return of(null)
      })
    ).subscribe(
      undefined,
      e => console.error(e),
      () => done()
    )
  })

  afterEach(done => mongoService.close().subscribe(() => done()))

  const omitVolatileUser = (user: User) => {
    const clone = omitVolatile(user)
    delete clone.services.password.bcrypt
    return clone
  }

  const createTestUsers$ = (displayNames: string[] = [
    'user-1',
    'user-2',
    'user-3',
    'user-4',
    'user-5',
    'admin',
    'superadmin'
  ]) => {
    return of(...displayNames).pipe(
      concatMap(name =>
        securityService.signUp({
          email: `${name}@test.local`,
          password: 'abc123'
        })),
      tap(user => expect(omitVolatileUser(user.user!)).toMatchSnapshot()),
      toArray(),
      map(users => ({
        user1: users[ 0 ].user!.id,
        user2: users[ 1 ].user!.id,
        user3: users[ 2 ].user!.id,
        user4: users[ 3 ].user!.id,
        user5: users[ 4 ].user!.id,
        admin: users[ 5 ].user!.id,
        superadmin: users[ 6 ].user!.id
      })),
      concatMap(({ user1, user2, user3, user4, user5, admin, superadmin }) => {
        return roleService.addUsersToRoles({
          userIds: [ user1, user2, user3, user4, user5 ],
          roles: [ 'members' ]
        }).pipe(
          concatMap(() => roleService.addUsersToRoles({
            userIds: [ user2 ],
            roles: [ 'users-two', 'admins' ],
            group: 'site-2'
          })),
          concatMap(() => roleService.addUsersToRoles({
            userIds: [ admin, superadmin ],
            roles: [ 'users', 'admins' ],
            group: 'site-5'
          })),
          concatMap(() => roleService.addUsersToRoles({
            userIds: [ superadmin ],
            roles: [ 'superadmins' ]
          })),
          mapTo({ user1, user2, user3, user4, user5, admin, superadmin })
        )
      })
    )
  }

  test('addUsersToRoles', done => {
    // addUsersToRoles() is called from createTestUsers$()
    createTestUsers$().pipe(
      concatMap(userIds => of(...Object.values(userIds))),
      concatMap(userId => securityService.user(userId)),
      tap(user => expect(omitVolatileUser(user!)).toMatchSnapshot())
    ).subscribe(
      undefined,
      undefined,
      () => {
        expect.assertions(14)
        done()
      }
    )
  })

  test('removeRole()', done => {
    const findRoles$ = roleService.findRoles({ project: { name: 1, _id: 0 } })

    createTestUsers$().pipe(
      // --- cannot remove roles that don't exist
      // --- cannot remove roles in use
      concatMap(() => findRoles$),
      tap(roles => expect(roles).toMatchSnapshot()),
      concatMap(() => of(...[ 'non-members', 'members' ])),
      concatMap(role => roleService.removeRole(role)),
      tap(removeRole => expect(removeRole).toBeFalsy()),
      toArray(),
      // --- can remove unused role
      concatMap(() => roleService.createRole('unused-role')),
      tap(id => expect(id.length).toBe(24)),
      concatMap(() => roleService.removeRole('unused-role')),
      tap(removeRole => expect(removeRole).toBeTruthy())
    ).subscribe(
      undefined,
      undefined,
      () => {
        expect.assertions(12)
        done()
      }
    )
  })

  test('getGroupsForUser()', done => {
    createTestUsers$().pipe(
      concatMap(userIds => of(...Object.values(userIds))),
      concatMap(userId => roleService.getGroupsForUser({ userId })),
      toArray(),
      tap(groupsForUser => expect(groupsForUser).toMatchSnapshot())
    ).subscribe(
      undefined,
      undefined,
      () => done()
    )
  })

  test('getGroupsForUser() with invalid user ID', done => {
    roleService.getGroupsForUser({ userId: '123456789012' })
      .subscribe(
        groups => expect(groups).toEqual([]),
        undefined,
        () => {
          expect.assertions(1)
          done()
        }
      )
  })

  test('getGroupsForUser() with invalid parameters', done => {
    createTestUsers$().pipe(
      // --- invalid user ID
      concatMap(userIds => {
        return roleService.getGroupsForUser({ userId: '123456789012' }).pipe(
          catchError(e => {
            expect(e).toBe('Error getting groups for user [123456789012]')
            return of(null)
          }),
          mapTo(userIds)
        )
      }),
      // --- user ID with no groups
      concatMap(userIds => {
        return roleService.getGroupsForUser({ userId: userIds.user1 }).pipe(
          tap(getGroupsForUser => expect(getGroupsForUser).toEqual([])),
          mapTo(userIds)
        )
      }),
      // --- user ID with group but not specifying role
      concatMap(userIds => {
        return roleService.getGroupsForUser({ userId: userIds.user2 }).pipe(
          tap(getGroupsForUser => expect(getGroupsForUser).toEqual([ 'site-2' ])),
          mapTo(userIds)
        )
      }),
      // --- user ID with group and specify role name that does not exist
      concatMap(userIds => {
        return roleService.getGroupsForUser({ userId: userIds.user2, role: 'no-such-role' }).pipe(
          tap(getGroupsForUser => expect(getGroupsForUser).toEqual([])),
          mapTo(userIds)
        )
      }),
      // --- user ID with group and specify role name that does exist but not assigned to this user
      concatMap(userIds => {
        return roleService.getGroupsForUser({ userId: userIds.user2, role: 'superadmins' }).pipe(
          tap(getGroupsForUser => expect(getGroupsForUser).toEqual([])),
          mapTo(userIds)
        )
      }),
      // --- user ID with group and specify correct role name
      concatMap(userIds => {
        return roleService.getGroupsForUser({ userId: userIds.user2, role: 'admins' }).pipe(
          tap(getGroupsForUser => expect(getGroupsForUser).toEqual([ 'site-2' ])),
          mapTo(userIds)
        )
      })
    ).subscribe(
      undefined,
      undefined,
      () => {
        expect.assertions(12)
        done()
      }
    )
  })

  test('findRoles()', done => {
    createTestUsers$().pipe(
      concatMap(() => roleService.findRoles({ project: { name: 1, _id: 0 } })),
      tap(roles => expect(roles).toMatchSnapshot())
    ).subscribe(
      undefined,
      undefined,
      () => {
        expect.assertions(8)
        done()
      }
    )
  })
})
