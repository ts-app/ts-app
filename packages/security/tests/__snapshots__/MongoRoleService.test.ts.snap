// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MongoRoleService addUsersToRoles 1`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-1@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-1",
    "email": "user-1@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService addUsersToRoles 2`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-2@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-2",
    "email": "user-2@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService addUsersToRoles 3`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-3@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-3",
    "email": "user-3@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService addUsersToRoles 4`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-4@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-4",
    "email": "user-4@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService addUsersToRoles 5`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-5@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-5",
    "email": "user-5@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService addUsersToRoles 6`] = `
Object {
  "emails": Array [
    Object {
      "email": "admin@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "admin",
    "email": "admin@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService addUsersToRoles 7`] = `
Object {
  "emails": Array [
    Object {
      "email": "superadmin@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "superadmin",
    "email": "superadmin@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService addUsersToRoles 8`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-1@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-1",
    "email": "user-1@test.local",
  },
  "roles": Array [
    Object {
      "group": "_global_",
      "role": "members",
    },
  ],
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService addUsersToRoles 9`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-2@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-2",
    "email": "user-2@test.local",
  },
  "roles": Array [
    Object {
      "group": "_global_",
      "role": "members",
    },
    Object {
      "group": "site-2",
      "role": "users-two",
    },
    Object {
      "group": "site-2",
      "role": "admins",
    },
    Object {
      "group": "site-5",
      "role": "users",
    },
    Object {
      "group": "site-5",
      "role": "admins",
    },
  ],
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService addUsersToRoles 10`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-3@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-3",
    "email": "user-3@test.local",
  },
  "roles": Array [
    Object {
      "group": "_global_",
      "role": "members",
    },
  ],
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService addUsersToRoles 11`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-4@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-4",
    "email": "user-4@test.local",
  },
  "roles": Array [
    Object {
      "group": "_global_",
      "role": "members",
    },
  ],
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService addUsersToRoles 12`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-5@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-5",
    "email": "user-5@test.local",
  },
  "roles": Array [
    Object {
      "group": "_global_",
      "role": "members",
    },
  ],
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService addUsersToRoles 13`] = `
Object {
  "emails": Array [
    Object {
      "email": "admin@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "admin",
    "email": "admin@test.local",
  },
  "roles": Array [
    Object {
      "group": "site-5",
      "role": "users",
    },
    Object {
      "group": "site-5",
      "role": "admins",
    },
  ],
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService addUsersToRoles 14`] = `
Object {
  "emails": Array [
    Object {
      "email": "superadmin@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "superadmin",
    "email": "superadmin@test.local",
  },
  "roles": Array [
    Object {
      "group": "site-5",
      "role": "users",
    },
    Object {
      "group": "site-5",
      "role": "admins",
    },
    Object {
      "group": "_global_",
      "role": "superadmins",
    },
  ],
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService findRoles() - simple, with query, with cursor & paging 1`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-1@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-1",
    "email": "user-1@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService findRoles() - simple, with query, with cursor & paging 2`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-2@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-2",
    "email": "user-2@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService findRoles() - simple, with query, with cursor & paging 3`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-3@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-3",
    "email": "user-3@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService findRoles() - simple, with query, with cursor & paging 4`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-4@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-4",
    "email": "user-4@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService findRoles() - simple, with query, with cursor & paging 5`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-5@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-5",
    "email": "user-5@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService findRoles() - simple, with query, with cursor & paging 6`] = `
Object {
  "emails": Array [
    Object {
      "email": "admin@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "admin",
    "email": "admin@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService findRoles() - simple, with query, with cursor & paging 7`] = `
Object {
  "emails": Array [
    Object {
      "email": "superadmin@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "superadmin",
    "email": "superadmin@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService findRoles() - simple, with query, with cursor & paging 8`] = `
Object {
  "docs": Array [
    Object {
      "name": "admins",
    },
    Object {
      "name": "members",
    },
    Object {
      "name": "superadmins",
    },
    Object {
      "name": "users",
    },
    Object {
      "name": "users-two",
    },
  ],
}
`;

exports[`MongoRoleService findRoles() - simple, with query, with cursor & paging 9`] = `
Object {
  "docs": Array [
    Object {
      "name": "admins",
    },
    Object {
      "name": "superadmins",
    },
  ],
}
`;

exports[`MongoRoleService findRoles() - simple, with query, with cursor & paging 10`] = `
Object {
  "docs": Array [
    Object {
      "name": "members",
    },
  ],
}
`;

exports[`MongoRoleService getGroupsForUser() 1`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-1@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-1",
    "email": "user-1@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getGroupsForUser() 2`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-2@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-2",
    "email": "user-2@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getGroupsForUser() 3`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-3@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-3",
    "email": "user-3@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getGroupsForUser() 4`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-4@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-4",
    "email": "user-4@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getGroupsForUser() 5`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-5@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-5",
    "email": "user-5@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getGroupsForUser() 6`] = `
Object {
  "emails": Array [
    Object {
      "email": "admin@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "admin",
    "email": "admin@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getGroupsForUser() 7`] = `
Object {
  "emails": Array [
    Object {
      "email": "superadmin@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "superadmin",
    "email": "superadmin@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getGroupsForUser() 8`] = `
Array [
  Array [],
  Array [
    "site-2",
    "site-5",
  ],
  Array [],
  Array [],
  Array [],
  Array [
    "site-5",
  ],
  Array [
    "site-5",
  ],
]
`;

exports[`MongoRoleService getGroupsForUser() with invalid parameters 1`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-1@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-1",
    "email": "user-1@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getGroupsForUser() with invalid parameters 2`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-2@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-2",
    "email": "user-2@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getGroupsForUser() with invalid parameters 3`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-3@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-3",
    "email": "user-3@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getGroupsForUser() with invalid parameters 4`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-4@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-4",
    "email": "user-4@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getGroupsForUser() with invalid parameters 5`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-5@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-5",
    "email": "user-5@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getGroupsForUser() with invalid parameters 6`] = `
Object {
  "emails": Array [
    Object {
      "email": "admin@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "admin",
    "email": "admin@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getGroupsForUser() with invalid parameters 7`] = `
Object {
  "emails": Array [
    Object {
      "email": "superadmin@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "superadmin",
    "email": "superadmin@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getRolesForUser() 1`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-1@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-1",
    "email": "user-1@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getRolesForUser() 2`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-2@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-2",
    "email": "user-2@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getRolesForUser() 3`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-3@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-3",
    "email": "user-3@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getRolesForUser() 4`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-4@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-4",
    "email": "user-4@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getRolesForUser() 5`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-5@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-5",
    "email": "user-5@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getRolesForUser() 6`] = `
Object {
  "emails": Array [
    Object {
      "email": "admin@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "admin",
    "email": "admin@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getRolesForUser() 7`] = `
Object {
  "emails": Array [
    Object {
      "email": "superadmin@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "superadmin",
    "email": "superadmin@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService getRolesForUser() 8`] = `
Array [
  Array [
    "members",
    "users-two",
    "admins",
    "users",
  ],
  Array [
    "members",
  ],
  Array [
    "members",
    "users-two",
    "admins",
  ],
  Array [
    "members",
    "users",
    "admins",
  ],
]
`;

exports[`MongoRoleService removeRole() 1`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-1@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-1",
    "email": "user-1@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService removeRole() 2`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-2@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-2",
    "email": "user-2@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService removeRole() 3`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-3@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-3",
    "email": "user-3@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService removeRole() 4`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-4@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-4",
    "email": "user-4@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService removeRole() 5`] = `
Object {
  "emails": Array [
    Object {
      "email": "user-5@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "user-5",
    "email": "user-5@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService removeRole() 6`] = `
Object {
  "emails": Array [
    Object {
      "email": "admin@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "admin",
    "email": "admin@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService removeRole() 7`] = `
Object {
  "emails": Array [
    Object {
      "email": "superadmin@test.local",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "superadmin",
    "email": "superadmin@test.local",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoRoleService removeRole() 8`] = `
Object {
  "docs": Array [
    Object {
      "name": "admins",
    },
    Object {
      "name": "members",
    },
    Object {
      "name": "superadmins",
    },
    Object {
      "name": "users",
    },
    Object {
      "name": "users-two",
    },
  ],
}
`;
